app/src/org/commcare/services/CommCareFirebaseMessagingService.java

```diff
@@ -0,0 +1,148 @@
+package org.commcare.services;
+
+import android.app.NotificationManager;
+import android.app.PendingIntent;
+import android.content.Intent;
+import android.os.Build;
+
+import androidx.core.app.NotificationCompat;
+
+import com.google.firebase.messaging.FirebaseMessagingService;
+import com.google.firebase.messaging.RemoteMessage;
+
+import org.commcare.CommCareNoficationManager;
+import org.commcare.activities.DispatchActivity;
+import org.commcare.dalvik.R;
+import org.commcare.util.LogTypes;
+import org.commcare.utils.FirebaseMessagingUtil;
+import org.javarosa.core.services.Logger;
+import org.joda.time.DateTime;
+
+import java.util.Map;
+
+/**
+ * This service responds to any events/messages from Firebase Cloud Messaging. The intention is to
+ * offer an entry point for any message from FCM and trigger the necessary steps based on the action
+ * key.
+ */
+public class CommCareFirebaseMessagingService extends FirebaseMessagingService {
+
+    private final static int FCM_NOTIFICATION = R.string.fcm_notification;
+    private enum ActionTypes{
+        SYNC,
+        INVALID
+    }
+
+   /**
+    * Upon receiving a new message from FCM, CommCare needs to:
+    * 1) Trigger the notification if the message contains a Notification object. Note that the
+    *    presence of a Notification object causes the onMessageReceived to not be called when the
+    *    app is in the background, which means that the data object won't be processed from here
+    * 2) Verify if the message contains a data object and trigger the necessary steps according
+    *    to the action it carries
+    *
+    */
+    @Override
+    public void onMessageReceived(RemoteMessage remoteMessage) {
+        Logger.log(LogTypes.TYPE_FCM, "Message received: " + remoteMessage.getMessageId());
+        Map<String, String> payloadData = remoteMessage.getData();
+        RemoteMessage.Notification payloadNotification = remoteMessage.getNotification();
```

Comment:
Can a notification contains both `data` and `notification` object ?

Reply:
bumping

Reply:
> I don't completely understand. How do we handle the sync notifcation when the app is in background ? Do we just discard the notification ? Or it's routed thorough a different pathway ?

Yeah, this is a bit tricky. So, when the app is in the background and the message contains a `notification` object, Firebase SDK takes care of delivering the notification to the _Notification Drawer_. In case the message also contains a `data` object, Firebase SDK adds it to the extras of the intent of the launcher activity of the app. This is the default behaviour, so when the user taps the notification it opens the app launcher with the `data` object as an extra. 

> Also when onMessageReceived gets called (app is in foreground), It seems like we can only have one of notification or data payload. 

We can have both, but in this case because we want to trigger the sync when the app is in the background too, we are having messages with only the `data` object

Reply:
> In case the message also contains a data object, Firebase SDK adds it to the extras of the intent of the launcher activity of the app. This is the default behaviour, so when the user taps the notification it opens the app launcher with the data object as an extra.

That means we should be handling these intents on app startup right ? Or they by default gets delivered to this messaging service when user clicks on notification and pull the app in foreground ?

Reply:
@shubham1g5 This is supposed to be handled on app startup but it's not yet implemented because our current implementation on HQ doesn't trigger FCM messages with both `notification` and `data` objects (see image below). So, the intention is to incorporate that on a later stage depending on the feedback we get as well. 
![image](https://github.com/dimagi/commcare-android/assets/19228119/f0da05c9-7621-4ca0-b744-06457389743f)

Reply:
Should not we still need to handle only `data` messages that gets send when app is in background ? if not, How does the app handle sync data messages that gets delivered to the phone when the CC is in background ? For ex, my current understanding is -  

1. We trigger a notification with Action 'Background Sync' and type data messages
2. This notification gets delivered to phone when CC is in background. Therefore Firebase adds the notification to the notification drawer ? (Not sure if it's true for only data messages)
3. On clicking notification, Firebase launches CC with the intent containing data action. But since we are not handling intents with data payload, background sync doesn't get triggered. 

Does it sound right to you ?

Reply:
When the `message` doesn't contain a `notification` object and the app is in the background, the message is delivered to the FCM Service, so the `onMessageDelivered` is called. More about this can be found [here](https://firebase.google.com/docs/cloud-messaging/android/receive#handling_messages).

Reply:
got it, thanks for confirming.